File: Source/printer.m
/* -*- c-basic-offset: 8 -*-
   rdesktop: A Remote Desktop Protocol client.
   Copyright (C) Matthew Chapman 1999-2005
   Rewritten for CoRD by Dorian Johnson, 2008

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2 of the License, or
   (at your option) any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software
   Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
*/

#import "CRDShared.h"
#import "rdesktop.h"
#import "Carbon/Carbon.h"
#import "sys/fcntl.h"

static NTStatus printer_create(RDConnectionRef conn, uint32 device_id, uint32 access, uint32 share_mode, uint32 disposition, uint32 flags, char *filename, NTHandle * handle);
static NTStatus printer_close(RDConnectionRef conn, NTHandle handle);
static NTStatus printer_write(RDConnectionRef conn, NTHandle handle, uint8 * data, uint32 length, uint32 offset, uint32 * result);


void
printer_enum_devices(RDConnectionRef conn)
{
	CFArrayRef printerList;
	
	PMServerCreatePrinterList(kPMServerLocal, &printerList);
	
	for (int i = 0; i < CFArrayGetCount(printerList); i++, conn->numDevices++)
	{
		PMPrinter printer = (void *)CFArrayGetValueAtIndex(printerList, i);
		const char *printerName = [[NSString stringWithString:(NSString *)PMPrinterGetName(printer)] cString];

		RDRedirectedDevice *device = &conn->rdpdrDevice[conn->numDevices];
		RDPrinterInfo *printerInfo = (RDPrinterInfo *) xmalloc(sizeof(RDPrinterInfo));
		
		printerInfo->printer = printer;
		
		printerInfo->rdpName = xmalloc(strlen(printerName)+1);
		strcpy(printerInfo->rdpName, printerName);
		
		strcpy(device->name, "PRN");
		strcat(device->name, l_to_a(i + 1, 10));
		
		const char *driverName = "HP Color LaserJet 8500 PS";
		printerInfo->rdpDriver = xmalloc(strlen(driverName)+1);
		strcpy(printerInfo->rdpDriver, driverName);
		
		device->device_type = DEVICE_TYPE_PRINTER;
		device->handle = 0;
		device->local_path = NULL;
		device->pdevice_data = (void *)printerInfo;
	}
}

static NTStatus
printer_create(RDConnectionRef conn, uint32 device_id, uint32 access, uint32 share_mode, uint32 disposition, uint32 flags, char *filename, NTHandle * handle)
{	
	RDRedirectedDevice *device = &conn->rdpdrDevice[device_id];
	const char *tempPath = [[CRDTemporaryFile() stringByAppendingString:@".eps"] cStringUsingEncoding:NSASCIIStringEncoding];
	
	device->local_path = xmalloc(strlen(tempPath)+1);
	strcpy(device->local_path, tempPath);
	
	int fd = open(tempPath, O_RDWR | O_TRUNC | O_APPEND | O_CREAT, S_IRWXU | S_IRGRP | S_IROTH);
	
	*handle = device->handle = fd;
	
	return STATUS_SUCCESS;
}

static NTStatus
printer_close(RDConnectionRef conn, NTHandle handle)
{
	RDRedirectedDevice *device = &conn->rdpdrDevice[get_device_index(conn, handle)];
	RDPrinterInfo *printerInfo = (void *)device->pdevice_data;
	
	// Close the temp file
	close(handle);
	
	// Create printing environment variables, send to printer
	PMPrintSession currentSession;
	PMCreateSession(&currentSession);
	
	PMPrintSettings printSettings;
	PMCreatePrintSettings(&printSettings);
	PMSessionDefaultPrintSettings(currentSession, printSettings);
	
	PMPageFormat pageFormat;
	PMCreatePageFormat(&pageFormat);
	PMSessionDefaultPageFormat(currentSession, pageFormat);
	
	CFURLRef filePath = CFURLCreateFromFileSystemRepresentation(NULL, (void *)device->local_path, strlen(device->local_path), false);
	
	OSStatus err =  PMPrinterPrintWithFile(printerInfo->printer, printSettings, pageFormat, CFStringCreateWithCString(NULL, "application/postscript", kCFStringEncodingASCII), filePath);

	NSLog(@"printer_close err=%d", err);

	PMRelease(currentSession);
	PMRelease(printSettings);
	PMRelease(pageFormat);
	CFRelease(filePath);
	
	// Delete the temp file
	remove(device->local_path);
	
	return STATUS_SUCCESS;
}

static NTStatus
printer_write(RDConnectionRef conn, NTHandle handle, uint8 * data, uint32 length, uint32 offset, uint32 * result)
{
	*result = length * write(handle, data, length);
	
	// xxx: need to catch errors
	
	*result = length;
	return STATUS_SUCCESS;
}

DEVICE_FNS printer_fns = {
	printer_create,
	printer_close,
	NULL,			/* read */
	printer_write,
	NULL			/* device_control */
};

File: Source/printer.c
#import "rdesktop.h"

static PRINTER *
get_printer_data(rdcConnection conn, NTHANDLE handle)
{
	int index;

	for (index = 0; index < RDPDR_MAX_DEVICES; index++)
	{
		if (handle == conn->rdpdrDevice[index].handle)
			return (PRINTER *) conn->rdpdrDevice[index].pdevice_data;
	}
	return NULL;
}

int
printer_enum_devices(rdcConnection conn, uint32 * id, char *optarg)
{
	PRINTER *pprinter_data;

	char *pos = optarg;
	char *pos2;
	int count = 0;
	int already = 0;

	/* we need to know how many printers we've already set up
	   supplied from other -r flags than this one. */
	while (count < *id)
	{
		if (conn->rdpdrDevice[count].device_type == DEVICE_TYPE_PRINTER)
			already++;
		count++;
	}

	count = 0;

	if (*optarg == ':')
		optarg++;

	while ((pos = next_arg(optarg, ',')) && *id < RDPDR_MAX_DEVICES)
	{
		pprinter_data = (PRINTER *) xmalloc(sizeof(PRINTER));

		strcpy(conn->rdpdrDevice[*id].name, "PRN");
		strcat(conn->rdpdrDevice[*id].name, l_to_a(already + count + 1, 10));

		/* first printer is set as default printer */
		if ((already + count) == 0)
			pprinter_data->default_printer = True;
		else
			pprinter_data->default_printer = False;

		pos2 = next_arg(optarg, '=');
		if (*optarg == (char) 0x00)
			pprinter_data->printer = "mydeskjet";	/* set default */
		else
		{
			pprinter_data->printer = xmalloc(strlen(optarg) + 1);
			strcpy(pprinter_data->printer, optarg);
		}

		if (!pos2 || (*pos2 == (char) 0x00))
			pprinter_data->driver = "HP Color LaserJet 8500 PS";	/* no printer driver supplied set default */
		else
		{
			pprinter_data->driver = xmalloc(strlen(pos2) + 1);
			strcpy(pprinter_data->driver, pos2);
		}

		printf("PRINTER %s to %s driver %s\n", conn->rdpdrDevice[*id].name,
		       pprinter_data->printer, pprinter_data->driver);
		conn->rdpdrDevice[*id].device_type = DEVICE_TYPE_PRINTER;
		conn->rdpdrDevice[*id].pdevice_data = (void *) pprinter_data;
		count++;
		(*id)++;

		optarg = pos;
	}
	return count;
}

static NTSTATUS
printer_create(rdcConnection conn, uint32 device_id, uint32 access, uint32 share_mode, uint32 disposition, uint32 flags,
	       char *filename, NTHANDLE * handle)
{
	char cmd[256];
	PRINTER *pprinter_data;

	pprinter_data = (PRINTER *) conn->rdpdrDevice[device_id].pdevice_data;

	/* default printer name use default printer queue as well in unix */
	if (pprinter_data->printer == "mydeskjet")
	{
		pprinter_data->printer_fp = popen("lpr", "w");
	}
	else
	{
		sprintf(cmd, "lpr -P %s", pprinter_data->printer);
		pprinter_data->printer_fp = popen(cmd, "w");
	}

	conn->rdpdrDevice[device_id].handle = fileno(pprinter_data->printer_fp);
	*handle = conn->rdpdrDevice[device_id].handle;
	return STATUS_SUCCESS;
}

static NTSTATUS
printer_close(rdcConnection conn, NTHANDLE handle)
{
	int i = get_device_index(conn, handle);
	if (i >= 0)
	{
		PRINTER *pprinter_data = conn->rdpdrDevice[i].pdevice_data;
		if (pprinter_data)
			pclose(pprinter_data->printer_fp);
		conn->rdpdrDevice[i].handle = 0;
	}
	return STATUS_SUCCESS;
}

static NTSTATUS
printer_write(rdcConnection conn, NTHANDLE handle, uint8 * data, uint32 length, uint32 offset, uint32 * result)
{
	PRINTER *pprinter_data;

	pprinter_data = get_printer_data(conn, handle);
	*result = length * fwrite(data, length, 1, pprinter_data->printer_fp);

	if (ferror(pprinter_data->printer_fp))
	{
		*result = 0;
		return STATUS_INVALID_HANDLE;
	}
	return STATUS_SUCCESS;
}

DEVICE_FNS printer_fns = {
	printer_create,
	printer_close,
	NULL,			/* read */
	printer_write,
	NULL			/* device_control */
};
